primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -910,4 +1152,77 @@ mod tests {
 
         Ok(())
     }
+
+    #[cfg(feature = "icicle")]
+    mod icicle {
+        use super::*;
+        use crate::{
+            icicle_deps::{curves::*, *},
+            pcs::univariate_kzg::icicle::GPUCommit,
+        };
+        use ark_ec::{short_weierstrass::Affine, CurveConfig};
+
+        fn test_gpu_e2e_template<E, C>() -> Result<(), PCSError>
+        where
+            C: IcicleCurve + MSM<C>,
+            C::ScalarField: ArkConvertible<ArkEquivalent = E::ScalarField>,
+            C::BaseField:
+                ArkConvertible<ArkEquivalent = <C::ArkSWConfig as CurveConfig>::BaseField>,
+            <C::ArkSWConfig as CurveConfig>::BaseField: PrimeField,
+            E: Pairing<G1Affine = Affine<<C as IcicleCurve>::ArkSWConfig>>,
+            UnivariateKzgPCS<E>: GPUCommit<E, C>,
+        {
+            let rng = &mut test_rng();
+            let stream = warmup_new_stream().unwrap();
+
+            let supported_degree = 2usize.pow(15);
+            let pp = UnivariateKzgPCS::<E>::gen_srs_for_testing(rng, supported_degree)?;
+            let (full_ck, _vk) = pp.trim(supported_degree)?;
+            let mut srs_on_gpu =
+                <UnivariateKzgPCS<E> as GPUCommit<E, C>>::load_prover_param_to_gpu(
+                    full_ck,
+                    supported_degree,
+                )?;
+
+            for _ in 0..10 {
+                let degree = usize::rand(rng) % 10000;
+                let (ck, vk) = pp.trim(degree)?;
+                let p =
+                    <DensePolynomial<E::ScalarField> as DenseUVPolynomial<E::ScalarField>>::rand(
+                        degree, rng,
+                    );
+                let comm = <UnivariateKzgPCS<E> as GPUCommit<E, C>>::commit_with_gpu(&ck, &p)?;
+                {
+                    // step-by-step commitment on gpu
+                    let poly_on_gpu =
+                        <UnivariateKzgPCS<E> as GPUCommit<E, C>>::load_poly_to_gpu(&p)?;
+                    let msm_result_on_gpu =
+                        <UnivariateKzgPCS<E> as GPUCommit<E, C>>::commit_on_gpu(
+                            &mut srs_on_gpu,
+                            &poly_on_gpu,
+                            &stream,
+                        )?;
+                    let comm2 = <UnivariateKzgPCS<E> as GPUCommit<E, C>>::load_commitment_to_host(
+                        msm_result_on_gpu,
+                        &stream,
+                    )?;
+                    assert_eq!(comm, comm2);
```

Comment:
Should we add a single test to compare the GPU results with the CPU results? This one could be a small instance

Reply:
i guess we could, but the fact that I'm passing this result to verify in `open()` and `verify()` is a testament that it is correct.

Reply:
done in [ca5b494](https://github.com/EspressoSystems/jellyfish/pull/498/commits/ca5b4942c0253dcda433abb2958f52e633cbae6e)